Fix #494: Reject a promise on error condition

nrchandan opened the pull request.

flovilmart commented on src/ParseObject.js:
```diff
@@ -1847,6 +1847,8 @@ var DefaultController = {
           }).then((response, status, xhr) => {
             batchReturned.resolve(response, status);
+          }, (error) => {
+            batchReturned.reject(new _ParseError2.default(undefined, error.message));
```
this seems wrong, should probably be `new ParseError(error.message);`

nrchandan commented:
Thanks for pointing it out. Will correct and push shortly.

On Oct 23, 2017 17:51, "Florent Vilmart" <notifications@github.com> wrote:

> *@flovilmart* commented on this pull request.
> ------------------------------
>
> In src/ParseObject.js
> <https://github.com/parse-community/Parse-SDK-JS/pull/495#discussion_r146246771>
> :
>
> > @@ -1847,6 +1847,8 @@ var DefaultController = {
>              }, options);
>            }).then((response, status, xhr) => {
>              batchReturned.resolve(response, status);
> +          }, (error) => {
> +            batchReturned.reject(new _ParseError2.default(undefined, error.message));
>
> this seems wrong, should probably be new ParseError(error.message);
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/parse-community/Parse-SDK-JS/pull/495#pullrequestreview-71159792>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABYXzgPktVZjE9eOTIzTOvJHTaCKEYx8ks5svIS5gaJpZM4QCuR6>
> .
>

flovilmart commented:
Also consider adding tests as you are changing a public interface, and this may be unexpected for many users.

On Oct 23, 2017, 08:29 -0400, Chandan Kumar <notifications@github.com>, wrote:
> Thanks for pointing it out. Will correct and push shortly.
>
> On Oct 23, 2017 17:51, "Florent Vilmart" <notifications@github.com> wrote:
>
> > *@flovilmart* commented on this pull request.
> > ------------------------------
> >
> > In src/ParseObject.js
> > <https://github.com/parse-community/Parse-SDK-JS/pull/495#discussion_r146246771>
> > :
> >
> > > @@ -1847,6 +1847,8 @@ var DefaultController = {
> > }, options);
> > }).then((response, status, xhr) => {
> > batchReturned.resolve(response, status);
> > + }, (error) => {
> > + batchReturned.reject(new _ParseError2.default(undefined, error.message));
> >
> > this seems wrong, should probably be new ParseError(error.message);
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/parse-community/Parse-SDK-JS/pull/495#pullrequestreview-71159792>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/ABYXzgPktVZjE9eOTIzTOvJHTaCKEYx8ks5svIS5gaJpZM4QCuR6>
> > .
> >
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

nrchandan commented:
Tried going through the tests in ParseObjectTest.js. Currently, there are no tests that cause failure of saveAll. I'm not sure how to make it fail.

nrchandan commented:
I think I can make it fail. Let me give it a try.

On Oct 23, 2017 18:03, "Florent Vilmart" <notifications@github.com> wrote:

> Also consider adding tests as you are changing a public interface, and
> this may be unexpected for many users.
>
> On Oct 23, 2017, 08:29 -0400, Chandan Kumar <notifications@github.com>,
> wrote:
> > Thanks for pointing it out. Will correct and push shortly.
> >
> > On Oct 23, 2017 17:51, "Florent Vilmart" <notifications@github.com>
> wrote:
> >
> > > *@flovilmart* commented on this pull request.
> > > ------------------------------
> > >
> > > In src/ParseObject.js
> > > <https://github.com/parse-community/Parse-SDK-JS/pull/
> 495#discussion_r146246771>
> > > :
> > >
> > > > @@ -1847,6 +1847,8 @@ var DefaultController = {
> > > }, options);
> > > }).then((response, status, xhr) => {
> > > batchReturned.resolve(response, status);
> > > + }, (error) => {
> > > + batchReturned.reject(new _ParseError2.default(undefined,
> error.message));
> > >
> > > this seems wrong, should probably be new ParseError(error.message);
> > >
> > > —
> > > You are receiving this because you authored the thread.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/parse-community/Parse-SDK-JS/pull/
> 495#pullrequestreview-71159792>,
> > > or mute the thread
> > > <https://github.com/notifications/unsubscribe-auth/
> ABYXzgPktVZjE9eOTIzTOvJHTaCKEYx8ks5svIS5gaJpZM4QCuR6>
> > > .
> > >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub, or mute the thread.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/parse-community/Parse-SDK-JS/pull/495#issuecomment-338643954>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABYXzju05NnnBPnCVwiv5IHIjH4SZgkAks5svIeYgaJpZM4QCuR6>
> .
>

flovilmart commented on integration/test/ParseObjectTest.js:
```diff
@@ -254,6 +254,14 @@ describe('Parse Object', () => {
+  it('cannot set an invalid date', (done) => {
+    let obj = new TestObject();
+    obj.set('when', new Date(Date.parse(null)));
+    Parse.Object.saveAll([obj]).fail((e) => {
```
Please add .then(done.fail)

flovilmart commented on integration/test/ParseObjectTest.js:
```diff
@@ -254,6 +254,14 @@ describe('Parse Object', () => {
+    let obj = new TestObject();
+    obj.set('when', new Date(Date.parse(null)));
+    Parse.Object.saveAll([obj]).fail((e) => {
+      done();
```
Please test for the expected error message.

flovilmart commented:
Also, the coverage is not improved as you’ve added an integration test and not a jest unit test. You may wanna consider doing that as there’s no way for now to ensure the code path is properly covered.

nrchandan commented:
That makes sense. Let me try adding a unit test.

On Oct 24, 2017 6:40 AM, "Florent Vilmart" <notifications@github.com> wrote:

> Also, the coverage is not improved as you’ve added an integration test and
> not a jest unit test. You may wanna consider doing that as there’s no way
> for now to ensure the code path is properly covered.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/parse-community/Parse-SDK-JS/pull/495#issuecomment-338841032>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABYXzkQ7fDhrJg42-bUJe1GOMY3dPSxLks5svTj7gaJpZM4QCuR6>
> .
>

nrchandan commented:
Found two duplicate unit tests (both in the unit test file `ParseObject-test.js`):
`can save an array of objects` is repeated twice (no diff)
`can save the object` appears twice with a negligible difference (objectId: 'P1' vs 'P5')

Should I remove these two in the same PR, or would you recommend creating a new PR for this?

flovilmart commented:
I prefer we don’t remove unit tests, PR’s should be as atomic as possible.

nrchandan commented:
Thanks Florent. Let me undo that and push again.

On Tue, Oct 24, 2017 at 5:01 PM, Florent Vilmart <notifications@github.com>
wrote:

> I prefer we don’t remove unit tests, PR’s should be as atomic as possible.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/parse-community/Parse-SDK-JS/pull/495#issuecomment-338960146>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABYXzlkmLxMjAPYIx06a6lbqqv9Sswb7ks5svcp6gaJpZM4QCuR6>
> .
>

nrchandan commented:
Added a new error code: 
```
/**
 * Error code indicating an invalid value of an object field is passed.
 * @property INVALID_VALUE
 * @static
 * @final
 */
ParseError.INVALID_VALUE = 603
```

Also, updated the tests to check for both error message and this new error code.

nrchandan commented:
Unit test has passed, but integration failed. Both the tests are very similar. @flovilmart Could you help me fix it, or would it make sense to remove it instead?

nrchandan commented:
Hi @flovilmart, I made a couple of changes. Could you take a look at this revised PR and let me know if this looks good for merging.

flovilmart commented:
This is looking good now! Let's go!
